File: backend/datum/services/technical_terms.py
"""Regex-based technical term extraction for exact-match search."""

from __future__ import annotations

from dataclasses import dataclass
import re

TECHNICAL_TERMS_PIPELINE_NAME = "regex-technical-terms"
TECHNICAL_TERMS_PIPELINE_VERSION = "regex-v1"

@dataclass(slots=True)
class TermMatch:
    normalized_text: str
    raw_text: str
    term_type: str
    start_char: int
    end_char: int
    confidence: float = 1.0


PATTERNS: list[tuple[str, re.Pattern[str], float]] = [
    (
        "api_route",
        re.compile(r"(?:GET|POST|PUT|DELETE|PATCH|HEAD|OPTIONS)\s+(/[\w./{}\-:]+)"),
        1.0,
    ),
    (
        "file_path",
        re.compile(r"(?<!\w)(\.\.?/[\w./-]+|/[\w./-]+(?:\.[\w.-]+)?)"),
        0.9,
    ),
    ("env_var", re.compile(r"\b([A-Z][A-Z0-9_]{2,})\b"), 0.8),
    ("version", re.compile(r"\bv?\d+\.\d+(?:\.\d+)?(?:-[\w.]+)?\b"), 0.9),
    (
        "sql_identifier",
        re.compile(r"(?:FROM|TABLE|JOIN|INTO|UPDATE|INDEX\s+ON)\s+([A-Za-z_][\w$]*)", re.IGNORECASE),
        0.9,
    ),
    (
        "package",
        re.compile(r"(?:pip\s+install|npm\s+install|yarn\s+add)\s+([\w\[\].-]+(?:\s+[\w\[\].-]+)*)", re.IGNORECASE),
        0.8,
    ),
    ("port", re.compile(r"(?::(\d{2,5})(?!\d))|(?:port\s+(\d{2,5}))", re.IGNORECASE), 0.7),
]

ENV_STOPWORDS = {
    "AND",
    "API",
    "CSS",
    "DELETE",
    "FROM",
    "GET",
    "HEAD",
    "HTML",
    "HTTP",
    "HTTPS",
    "INTO",
    "JOIN",
    "JSON",
    "NOT",
    "OPTIONS",
    "PATCH",
    "POST",
    "PUT",
    "SELECT",
    "SQL",
    "TABLE",
    "THE",
    "URL",
    "WHERE",
    "YAML",
}


def extract_technical_terms(text: str) -> list[TermMatch]:
    if not text:
        return []

    matches: list[TermMatch] = []
    seen_spans: set[tuple[int, int]] = set()

    for term_type, pattern, confidence in PATTERNS:
        for match in pattern.finditer(text):
            groups = [group for group in match.groups() if group is not None]

            if term_type == "package" and groups:
                package_group = groups[0]
                group_start = match.start() + match.group().index(package_group)
                cursor = group_start
                for package in package_group.split():
                    start = text.find(package, cursor)
                    if start < 0:
                        continue
                    end = start + len(package)
                    cursor = end
                    if (start, end) in seen_spans:
                        continue
                    seen_spans.add((start, end))
                    matches.append(
                        TermMatch(
                            normalized_text=package.lower().split("[", 1)[0],
                            raw_text=package,
                            term_type=term_type,
                            start_char=start,
                            end_char=end,
                            confidence=confidence,
                        )
                    )
                continue

            if groups:
                raw = groups[0]
                start = match.start() + match.group().index(raw)
                end = start + len(raw)
            else:
                raw = match.group()
                start = match.start()
                end = match.end()

            if (start, end) in seen_spans:
                continue
            if term_type == "env_var" and raw in ENV_STOPWORDS:
                continue

            seen_spans.add((start, end))
            matches.append(
                TermMatch(
                    normalized_text=_normalize_term(term_type, raw),
                    raw_text=raw,
                    term_type=term_type,
                    start_char=start,
                    end_char=end,
                    confidence=confidence,
                )
            )

    return sorted(matches, key=lambda item: (item.start_char, item.end_char))


def _normalize_term(term_type: str, raw_text: str) -> str:
    if term_type in {"package", "sql_identifier"}:
        return raw_text.lower()
    return raw_text

File: frontend/src/lib/technical-terms.ts
export interface TechnicalTerm {
  normalizedText: string
  rawText: string
  termType: string
  startChar: number
  endChar: number
  confidence: number
}

interface HeadingItem {
  id: string
  level: number
  text: string
}

const technicalTermPatterns: Array<{ termType: string; pattern: RegExp; confidence: number }> = [
  {
    termType: 'api_route',
    pattern: /(?:GET|POST|PUT|DELETE|PATCH|HEAD|OPTIONS)\s+(\/[\w./{}\-:]+)/g,
    confidence: 1,
  },
  {
    termType: 'file_path',
    pattern: /(?<!\w)(\.\.?\/[\w./-]+|\/[\w./-]+(?:\.[\w.-]+)?)/g,
    confidence: 0.9,
  },
  {
    termType: 'env_var',
    pattern: /\b([A-Z][A-Z0-9_]{2,})\b/g,
    confidence: 0.8,
  },
  {
    termType: 'version',
    pattern: /\bv?\d+\.\d+(?:\.\d+)?(?:-[\w.]+)?\b/g,
    confidence: 0.9,
  },
  {
    termType: 'sql_identifier',
    pattern: /(?:FROM|TABLE|JOIN|INTO|UPDATE|INDEX\s+ON)\s+([A-Za-z_][\w$]*)/gi,
    confidence: 0.9,
  },
  {
    termType: 'package',
    pattern: /(?:pip\s+install|npm\s+install|yarn\s+add)\s+([\w\[\].-]+(?:\s+[\w\[\].-]+)*)/gi,
    confidence: 0.8,
  },
  {
    termType: 'port',
    pattern: /(?::(\d{2,5})(?!\d))|(?:port\s+(\d{2,5}))/gi,
    confidence: 0.7,
  },
]

const envStopwords = new Set([
  'AND',
  'API',
  'CSS',
  'DELETE',
  'FROM',
  'GET',
  'HEAD',
  'HTML',
  'HTTP',
  'HTTPS',
  'INTO',
  'JOIN',
  'JSON',
  'NOT',
  'OPTIONS',
  'PATCH',
  'POST',
  'PUT',
  'SELECT',
  'SQL',
  'TABLE',
  'THE',
  'URL',
  'WHERE',
  'YAML',
])

export function stripFrontmatter(content: string): string {
  return content.replace(/^---[\s\S]*?---\n*/, '')
}

export function slugifyHeading(value: string): string {
  return value
    .toLowerCase()
    .replace(/[`*_~]/g, '')
    .replace(/[^\w\s-]/g, '')
    .trim()
    .replace(/\s+/g, '-')
}

export function extractHeadings(markdown: string): HeadingItem[] {
  const headings: HeadingItem[] = []
  const headingCounts = new Map<string, number>()
  let inFence = false

  for (const line of stripFrontmatter(markdown).split('\n')) {
    if (line.trimStart().startsWith('```')) {
      inFence = !inFence
      continue
    }
    if (inFence) {
      continue
    }

    const match = /^(#{1,6})\s+(.+?)\s*$/.exec(line)
    if (!match) {
      continue
    }
    const level = match[1].length
    const text = match[2].replace(/\s+#*$/, '')
    const baseId = slugifyHeading(text)
    const duplicateCount = headingCounts.get(baseId) ?? 0
    headingCounts.set(baseId, duplicateCount + 1)
    headings.push({
      id: duplicateCount === 0 ? baseId : `${baseId}-${duplicateCount + 1}`,
      level,
      text,
    })
  }

  return headings
}

export function extractTechnicalTerms(text: string): TechnicalTerm[] {
  if (!text) {
    return []
  }

  const matches: TechnicalTerm[] = []
  const seenSpans = new Set<string>()

  for (const { termType, pattern, confidence } of technicalTermPatterns) {
    for (const match of text.matchAll(pattern)) {
      const groups = match.slice(1).filter(Boolean)

      if (termType === 'package' && groups[0]) {
        const packageGroup = groups[0]
        let cursor = text.indexOf(packageGroup, match.index ?? 0)
        for (const pkg of packageGroup.split(/\s+/)) {
          const start = text.indexOf(pkg, cursor)
          if (start < 0) {
            continue
          }
          const end = start + pkg.length
          cursor = end
          const spanKey = `${start}:${end}`
          if (seenSpans.has(spanKey)) {
            continue
          }
          seenSpans.add(spanKey)
          matches.push({
            normalizedText: pkg.toLowerCase().split('[', 1)[0],
            rawText: pkg,
            termType,
            startChar: start,
            endChar: end,
            confidence,
          })
        }
        continue
      }

      const rawText = groups[0] ?? match[0]
      if (!rawText) {
        continue
      }

      const startOffset = match[0].indexOf(rawText)
      const startChar = (match.index ?? 0) + Math.max(startOffset, 0)
      const endChar = startChar + rawText.length
      const spanKey = `${startChar}:${endChar}`

      if (seenSpans.has(spanKey)) {
        continue
      }
      if (termType === 'env_var' && envStopwords.has(rawText)) {
        continue
      }

      seenSpans.add(spanKey)
      matches.push({
        normalizedText:
          termType === 'package' || termType === 'sql_identifier'
            ? rawText.toLowerCase()
            : rawText,
        rawText,
        termType,
        startChar,
        endChar,
        confidence,
      })
    }
  }

  return matches.sort((left, right) => left.startChar - right.startChar || left.endChar - right.endChar)
}

export function uniqueTechnicalTerms(terms: TechnicalTerm[], limit = 24): TechnicalTerm[] {
  const unique = new Map<string, TechnicalTerm>()
  for (const term of terms) {
    const key = `${term.termType}:${term.rawText}`
    if (!unique.has(key)) {
      unique.set(key, term)
    }
  }

  return [...unique.values()]
    .sort((left, right) => right.rawText.length - left.rawText.length || left.rawText.localeCompare(right.rawText))
    .slice(0, limit)
}
